te-app/src/main/java/titanicsend/pattern/glengine/GLShader.java

```
@@ -211,127 +218,81 @@ public static ByteBuffer allocateMappedBuffer(int width, int height) {
-    this.mappedBufferWidth = width;
-    this.mappedBufferHeight = height;
-    this.mappedBuffer = buffer;
-    this.useMappedBuffer = true;
+  // Setup
+
+  protected GLShader addUniformSource(UniformSource uniformSource) {
+    if (uniformSource == null) {
+      throw new IllegalArgumentException("UniformSource cannot be null");
+    }
+    this.uniformSources.add(uniformSource);
+    return this;
   }
 
-  /** Activate this shader for rendering in the current context */
-  public void useProgram() {
-    gl4.glUseProgram(shaderProgram.getProgramId());
+  // get the active GL4 object for this shader.
+  public GL4 getGL4() {
+    return gl4;
   }
 
-  // initialization that can be done before the OpenGL context is available
-  public void createShaderProgram(FragmentShader fragmentShader, int xResolution, int yResolution) {
-    this.xResolution = xResolution;
-    this.yResolution = yResolution;
-    this.fragmentShader = fragmentShader;
-    this.vertexBuffer = Buffers.newDirectFloatBuffer(VERTICES.length);
-    this.indexBuffer = Buffers.newDirectIntBuffer(INDICES.length);
-    this.vertexBuffer.put(VERTICES);
-    this.indexBuffer.put(INDICES);
+  // Initialization
+
+  public void onActive() {
+    if (!initialized) {
+      init();
+    }
+  }
 
-    // allocate default buffer for reading offscreen surface to cpu memory
-    if (this.backBuffer == null) this.backBuffer = allocateBackBuffer();
+  public void onInactive() {
+    // TODO - anything we can temporarily release here?
+    // for the moment, we're leaving the shader program intact
+    // so it is very fast to reactivate.
+  }
+
+  public boolean isInitialized() {
+    return this.initialized;
   }
 
-  // Shader initialization that requires the OpenGL context
-  // Normally called each time the pattern is activated or
-  // when the effect is enabled.
-  public void init() {
+  /** Shader initialization that requires the OpenGL context. Call once. */
+  public final void init() {
```

related to above comment about synchronizing this check - another lightweight approach might be

```java
public final void init() {
  if (!this.initialized) {
    // ...
    this.initialized = true;
  }
}
```

---

Given it's single-threaded, style-wise I prefer setting `initialized` to true immediately after the check.  It groups the flag behavior all in one place which is fast to read on a scan, also removes any question about whether a partial initialization state could happen from entering the init loop but failing part way through.

And I prefer explicitly throwing an error if already initialized, rather than silently skip.  Initialization is a single-time action and if someone attempts it again that's a bug and should throw an error.  I added the `isInitialized()` method which can be checked to avoid this.